nautobot/docs/development/best-practices.md

```diff
@@ -63,3 +63,193 @@ class ExampleModel(PrimaryModel):
+False
+(Pdb) field.lookup_expr
+'exact' 
+```
+
+This means that the arguments for the field are being completely ignored and the hard-coded queryset `queryset.exclude(consoleports__isnull=value)` is all that is being run when this method is called. This hard-coding is impossible to introspect and therefore reverse.
+
+In fact, this method is completely unnecessary, because it could be replaced with options on the filter field itself:
+
+```python
+    console_ports = django_filters.BooleanFilter(
+        field_name="consoleports",  # The actual related field name
+        exclude=True,               # Perform an `.exclude()` vs. `.filter()``
+        lookup_expr="isnull",       # Perform `isnull` vs. `exact``
+        label="Has console ports",
+    )
+```
+
+Now, if we use another breakpoint (this time inside of the `DeviceListView.get()` since there’s no method we can break into), you can see that the filter field now has the correect attributes that can be used to accurately reverse this query:
+
+```python
+(Pdb) filterset = self.filterset(request.GET, self.queryset)
+(Pdb) field = filterset.filters["console_ports"]
+(Pdb) field.exclude
+True
+(Pdb) field.lookup_expr
+'isnull'
+```
+
+#### Generating Reversible Q Objects
+
+These field values could be used to construct a `Q()` query that looks something like:
+
+```python
+def generate_query(self, field_name, lookup_expr, value):
+    query = Q()
+    predicate = {f"{field_name}__{lookup_expr}": value}
+    if field.exclude:
+        query |= ~Q(**predicate)
+    else:
+        query |= Q(**predicate)
+    return query
+    
+## Somewhere else in business logic:
+field = filterset.filters[name]
+value = filterset.data[name]
+query = generate_query(field.field_name, field.lookup_expr, value)
+filterset.qs.filter(query).count()  # 339
+```
+
+But furthermore, this also becomes unnecessary for the common case, because we can **just trust the filterset itself to do the proper filtering and completely obviate the need to reverse the query at all.**
+
+Consider this input filter:
+
+```python
+filter_params = {"console_ports": True}
+```
+
+With properly configured filter fields, this will just work:
+
+```python
+DeviceFilterSet(filter_params).qs.count()  # 339
+```
+
+### Summary
+
+- For the vast majority of cases where we have method filters, it’s for Boolean filters
+- I am asserting for the common case **method filters are unnecessary technical debt and should be eliminated where better suited by proper use of filter field arguments**
+- Reversibility may not necessarily be required, but by properly defining `field_name`, `lookup_expr`, and `exclude` on filter fields, **introspection becomes deterministic and reversible queries can be reliably generated as needed.**
+- For exceptions such as `DeviceFilterSet.has_primary_ip` where it checks for both `Device.primary_ip4` OR `Device.primary_ip6`, method filters may still be necessary, however, they would be **the exception and not the norm.**
```

Comment:
I'd love to see this one fleshed out more - here we have a filter field that *does* need to use a `method` - so is it currently implemented as optimally and correctly as possible? Does it still pose problems, and if so what are they?

Reply:
This seems to also be answered by the same comment above. However, could still refine this into the pros/cons of using method filters and best practices for IF you do vs. why you shouldn't or try to avoid it?